pkg/manager/runtime.go

```diff
@@ -448,31 +448,44 @@ func (v *VirtletRuntimeService) UpdateRuntimeConfig(context.Context, *kubeapi.Up
 // UpdateContainerResources stores in domain on libvirt info about Cpuset
 // for container then looks for running emulator and tries to adjust its
 // current settings through cgroups
+// the config.ResourceUpdateInProgress provides a simple locking mechanism here instead of
+// adding a pod-containerID based map here to track containers being updated
 func (v *VirtletRuntimeService) UpdateContainerResources(ctx context.Context, req *kubeapi.UpdateContainerResourcesRequest) (*kubeapi.UpdateContainerResourcesResponse, error) {
 	glog.V(4).Infof("Update Container Resources : %v", req)
-	lcr := req.GetLinux()
-	if lcr == nil {
-		glog.Errorf("linuxContainerResource is not set in UpdateContainerResourceRequest")
-		return &kubeapi.UpdateContainerResourcesResponse{}, nil
-	}
-
-	lr := linuxContinerResourceToLinuxResource(lcr)
 
 	containerId := req.ContainerId
 	info, err := v.virtTool.ContainerInfo(containerId)
 	if err != nil {
 		return nil, err
 	}
 
+	// if an resource being updated currently, just ignore the call, kubelet will retry periodically on the request
+	if info.Config.ResourceUpdateInProgress == true {
+		glog.V(4).Infof("VM resource update in progress. Ignore the new request")
+		return &kubeapi.UpdateContainerResourcesResponse{}, nil
+	}
+
+	glog.V(4).Infof("Update VM config to indicate update resource is in progress")
+	v.virtTool.SetUpdateResourceUpdateInProgress(info.Id, true)
+
+	lcr := req.GetLinux()
+	if lcr == nil {
+		glog.Errorf("linuxContainerResource is not set in UpdateContainerResourceRequest")
+		return &kubeapi.UpdateContainerResourcesResponse{}, nil
+	}
+
+	lr := linuxContinerResourceToLinuxResource(lcr)
 	err = cgroups.UpdateVmCgroup(path.Join(info.Config.CgroupParent, containerId), lr)
 	if err != nil {
+		v.virtTool.SetUpdateResourceUpdateInProgress(info.Id, false)
 		return nil, err
 	}
 
 	// TODO: should revert the CG update if the update domain resource function failed
 	err = v.virtTool.UpdateDomainResources(containerId, lcr)
 	if err != nil {
 		glog.V(4).Infof("Update Domain Resource failed with error: %v", err)
+		v.virtTool.SetUpdateResourceUpdateInProgress(info.Id, false)
 		return nil, err
 	}
 
```

Comment:
Shouldn’t you set inProgress to false before you return in the success path?

Reply:
for the success paths, the flag is set in the callback events for memory changes. however, for cpu changes, it will be needed. will add code there.

Reply:
fixed in cpu update cases. setvcpus() is a synchronized call so it is a bit easier to reset there.

Reply:
Usually we keep lock/unlock pair in a same method. Separating them in different methods make it very difficult to track if unlock and unlock is paired.  Suggest refining the design to keep them together.

Reply:
I agree. this special case the lock has to be reset in the libvirt callback and the same time, we want to keep this async pattern in runtime. i will file a design bug to track this improvement.